Permission level degradation proposal (#66)
kubermatic/bulward: docs/proposals/permission-level-degradation.md

# Bulward Permission Level Degradation

> This document extends on the initial Bulward concept to enable degradation of user defined roles.

Reviewer: ```suggestion
This document extends on the initial Bulward concept to enable user defined roles scope reduction.
```

not sure if the degradation is the right word in this context.

## Background

> After installing bulward organization owners will have very limited permissions. Essentially they will only be allowed to interact with the Bulward API -> e.g. creating new Projects, organize users and create sub-roles.

Reviewer: ```suggestion
After installing bulward organization owners have limited permissions. Essentially they are allowed to interact with the Bulward API -> e.g. creating new Projects, organize users, and create sub-roles.
```

In order to actually grant an organization rights to do something, these rights need to be explicitly assigned to an Organization.
New Operators may ship `OrganizationRoleTemplates` to setup default roles to interact with it.

This is all fine, but it gets complicated when Organization Owners define their own sub-roles.

While granting new permissions is uncomplicated, removing permissions is less easy, because new roles utilizing these permissions might have been created.
This concept deals with the necessary permission/role cleanup, by introducing new theoretical concepts and extending the KubeCarrier API.

## Concepts

### Max Permission Level (Organization/Project)

> Organizations and Projects have both a maximum permission level that cannot be exceeded within.

Reviewer: How is this maximum permission configured? In the `Organization` / `Project` resource or?

> This maximum permission level results from all roles that are assigned to the organization unit and thus to the Organization/Project Owner.

Reviewer: This is tad underspecified. `All roles` --> do we mean `OrganizationRoleTemplate` roles or?

Organization/Project Owners may create sub-roles that don't exceed this maximum permission level.

> If roles are removed from the organization the maximum permission level drops and sub-roles need to be limited in order not to exceed this permission level.

Reviewer: Same, `roles` is referring to which resource exactly?

This way we can keep the user specification intact, while enabling only parts of his specification when it violates the max permission level.
We can also report detailed 

### New CRDs

## ProjectRole

`ProjectRole` objects control a `Role` object within the `Project` namespace, that the object was created in.
In difference to Kubernetes `Role` objects, Bulward will only consider RBAC rules in the definition that are not in violation of the max permission level of the project.

The accepted rules will be listed in the objects status to inform the user.

`ProjectRoles` are validated on creation and update to enforce the same prerequisites as Kubernetes:
https://kubernetes.io/docs/reference/access-authn-authz/rbac/#restrictions-on-role-creation-or-update

```yaml
apiVersion: bulward.io/v1alpha1
kind: ProjectRole
metadata:
  name: mycoolapps-admin
  namespace: project-namespace
rules:
- apiGroups:
  - my-corp.com
  resources:
  - mycoolapps
  verbs:
  - get
  - list
  - watch
  - create
  - update
  - patch
  - delete
status:
  phase: Established
  observedGeneration: 1
  conditions: []
  acceptedRules:
  - apiGroups:
    - my-corp.com
    resources:
    - mycoolapps

>     verbs: # limited verbs, due to a max permission issue

Reviewer 2: This means, if user tries to create ProjectRole which escalates the user's permission or max permission, we will still let the `ProjectRole` be created but with limited permissions? Why don't we forbidden to create this ProjectRole?

Author: I should probably point this out more prominently in the design document:

> `OrganizationRoles` are validated on creation and update to enforce the same prerequisites as Kubernetes:
> https://kubernetes.io/docs/reference/access-authn-authz/rbac/#restrictions-on-role-creation-or-update

Same for `ProjectRoles`. `*RoleBindings` are also validated the similar, but the link points to the `*Binding` specific validation rules.

Those checks include escalation checks, just follow the link.

Reviewer 2: I am sorry that I still don't get it :) 
For example, if User only has permission to do X but not Y, and User creates a `Role` with X and Y, Kubernetes API server will reject the API call with a forbidden error.
In our case, if User create a `ProjectRole` with X and Y, we will let it be created, but we only give people permission X (In the status, only X will be granted).
My question was why don't we also reject the call with a forbidden error when user creates `ProjectRole`? Why we let it be created instead?
Thanks!

Reviewer 2: Oh, now I see it. In your example, it is the case that the max permission is limited after the `ProjectRole` is created, in this case, we have the `acceptedRules` in the status to show User, that your permission is also limited.

    - get
    - list
    - watch
    - update
    - patch
```

## OrganizationRole

`OrganizationRole` objects control a `Role` object within the `Organization` namespace and all `Project` namespaces existing within the organization.
In difference to Kubernetes `Role` objects, Bulward will only consider RBAC rules in the definition that are not in violation of the max permission level of the organization.

The accepted rules will be listed in the objects status to inform the user.

`OrganizationRoles` are validated on creation and update to enforce the same prerequisites as Kubernetes:
https://kubernetes.io/docs/reference/access-authn-authz/rbac/#restrictions-on-role-creation-or-update

```yaml
apiVersion: bulward.io/v1alpha1
kind: OrganizationRole
metadata:
  name: mycoolapps-admin
rules:
- apiGroups:
  - my-corp.com
  resources:
  - mycoolapps
  verbs:
  - get
  - list
  - watch
  - create
  - update
  - patch
  - delete
status:
  phase: Established
  observedGeneration: 1
  conditions: []
  acceptedRules:
  - apiGroups:
    - my-corp.com
    resources:
    - mycoolapps
    verbs:
    - get
    - list
    - watch
    - update
    - patch
```

> ## ProjectRoleBinding

Reviewer: Why is this needed? Why ordinary rolebinding aren't enough?

Author: Because we want custom validation, at least to make sure that `ProjectRoleBinding` objects can only reference `ProjectRoles` and `OrganizationRoles` and `OrganizationRoleBindings` can only reference `OrganizationRoles`.

We don't want someone to try to bind directly to `Roles`/`ClusterRoles`

```yaml
apiVersion: bulward.io/v1alpha1
kind: ProjectRoleBinding
metadata:
  name: mycoolapps-admin
roleRef:
  kind: ProjectRole # or OrganizationRole
  apiGroup: bulward.io
  name: mycoolapps-admin
subjects:
- kind: User
  apiGroup: rbac.authorization.k8s.io
  name: sebastian@kubermatic.com
status:
  phase: Established
  observedGeneration: 1
  conditions: []
```

## OrganizationRoleBinding

`OrganizationRoleBinding` binds a `OrganizationRole` to a list of RBAC subjects across all projects and the organization namespace.
The only valid `roleRef` type is the `OrganizationRole` from the `bulward.io` api.

`OrganizationRoleBinding` objects are validated on creation and update to enforce the same prerequisites as Kubernetes:
https://kubernetes.io/docs/reference/access-authn-authz/rbac/#restrictions-on-role-binding-creation-or-update

```yaml
apiVersion: bulward.io/v1alpha1
kind: OrganizationRoleBinding
metadata:
  name: mycoolapps-admin
roleRef:
  kind: OrganizationRole
  apiGroup: bulward.io
  name: mycoolapps-admin
subjects:
- kind: User
  apiGroup: rbac.authorization.k8s.io
  name: sebastian@kubermatic.com
status:
  phase: Established
  observedGeneration: 1
  conditions: []
```

> ## Changes from initial concept

Reviewer: update the concept in the PR as well, to have it up to date.
